### PR TITLE
Simplify code around strerror_r

### PR DESCRIPTION
**Pull request by piponazo**
I am spending some time trying to clean all the warnings I get on gcc on Release mode. In this PR I just wanted to fix the warning I was getting by the usage of **strerror_r**:

```
/media/linuxDev/programming/exiv2/src/futils.cpp:349: warning: ignoring return value of ‘char* strerror_r(int, char*, size_t)’, declared with attribute warn_unused_result [-Wunused-result]
         strerror_r(error, buf, n);
         ~~~~~~~~~~^~~~~~~~~~~~~~~
```
The first strange thing I noticed is that the code was exercising the line 349 when I was expecting the code flow to reach the first block of code here:

```
# if defined EXV_STRERROR_R_CHAR_P && defined _GNU_SOURCE
        char *buf = 0;
        char buf2[n];
        std::memset(buf2, 0x0, n);
        buf = strerror_r(error, buf2, n);
# else
        char buf[n];
        std::memset(buf, 0x0, n);
        strerror_r(error, buf, n);
# endif
```
The reason is that we were not handling the definition of `EXV_STRERROR_R_CHAR_P` in none of our configuration systems (autotools, msvc solutions, cmake) and the logic in the C++ was very convoluted and difficult to understand. 

I started to investigate more about all the STRERROR stuff and I noticed that we had several definitions for controlling the usage of the strerror functions family. In this PR I am trying to simplify it a bit. These are the things I have done:

- Assume the usage of the **char *** version of `strerror_e` for linux. 
- Assume the usage of the **int** version of `strerror_e` for `__APPLE__` (XSI-Compliant). 
- Always assume the existence of `strerror` (It should be always available on C++98 compilers). We do not need to check whether it exists.

After few iterations I achieved to have everything working on CI (note that we already have some unit tests exercising that function on CI).

### CODE REVIEW

File: config/generateConfigFile.cmake
Comment (piponazo) [inline config/generateConfigFile.cmake]: This was always evaluating to true on Linux and Mac (false on windows). It did not make sense to use it.

**Review by clanmills — APPROVED** [review]
The only thing I can remember about this is that it caused pain.  If you've thought of a simpler/cleaner way to deal with this, I'm very happy to approve your changes.  If there are consequences for autotools and/or msvc, I'll investigate and suggest appropriate changes.

**Comment by D4N** [comment]
Is there a particular reason for using the gnu version of `strerror_r`? Otherwise I would suggest to switch to the XSI compliant version (the one used on OSX), since it is more portable and thread safe (the gnu version makes no thread safety guarantees). That should not be a huge problem, since we are nowhere defining `_GNU_SOURCE` and `_POSIX_C_SOURCE`.

The following define before the first include of `string.h` will ensure that the XSI compliant version is used with glibc:
``` c++
#define _POSIX_C_SOURCE 200112L
#ifdef _GNU_SOURCE
#undef _GNU_SOURCE
#endif
```
In theory, we could then drop the `#ifdef __APPLE__`.

Otherwise I am all in favor of this! There is really no point in checking whether the standard lib supports the functions required by the standard (if it doesn't, well then maybe switch to a better standard lib, eh?).

**Comment by clanmills** [comment]
@D4N I believe I inherited this horror without knowing what it did.  Duct-tape and glue was used to get it to build.

It would be good to "step back" and clean this up.  If C++11 gets us out of this mess, that's another reason to drop support for C++98.

**Comment by piponazo** [comment]
Good point @D4N , I will try to completely drop the GNU version usage and use always the XSI compliant verison.

**Comment by D4N** [comment]
@clanmills C++11 won't help any more than C++98 would, as strerror is required by C89 and C99. So if a compiler does not provide it, it is seriously broken (which of course does not mean it doesn't exist ;-) ).

@piponazo Thanks for your work! I think it should come down to just using the part from the __APPLE__ ifdef (plus the feature test macros in the header).

**Comment by piponazo** [comment]
@D4N I tried to do what you proposed but it does not work. It seems that on Linux, `_GNU_SOURCE` is defined when you use **libstdc++**:

http://gcc.gnu.org/onlinedocs/libstdc++/faq.html#faq.predefined

Therefore we need to use the GNU version of linux and the XSI on Mac. However I can replace the check of `__APPLE__` for `_GNU_SOURCE`

File: config/generateConfigFile.cmake
Comment (D4N) [inline config/generateConfigFile.cmake]: Why are you checking whether `strchr` exists? That is required by C89 & C99.

(I am just puzzled, since it's not used in this file.)

Comment (piponazo) [reply config/generateConfigFile.cmake]: Oh, damn it. I had a conflict when I rebased in top of master after I merged #294 , and I did not solve it correctly. Good catch! I will fix it and push the changes again.

Comment (piponazo) [reply config/generateConfigFile.cmake]: Fixed. I will merge once the CI do its job ;)

**Pull request merged**
